### ISSUE TITLE
Proposal: Lease Based Implementation

### ISSUE THREAD

**Issue by MrMDavidson** [rebus-org/Rebus.SqlServer#13]
Just wanted run an idea past you; how do you feel about a lease based implementation in SqlServer? The current implementation of SqlServer is basically;
 1. Start transaction
 2. Delete first message and return result
 3. Run handlers
 4. Commit transaction
If the handlers run for a long period of time you can run into issues with the transaction timing out resulting in the work being wasted. There's also the issue with transactions then potentially being promoted to distributed transactions and the overhead they cause.

I was considering an implementation where we introduce a new column to the `RebusMessage` table; `LeasedUntilUtc`. This implementation would then run as;
 1. Update the first row with a null `LeasedUntilUtc` to have a `LeasedUntilUtc` of some time in the future (say, 45s)
 2. Run handlers
 3. `ITransactionContext`'s `Commit()` would then be changed to delete the row we were dealing with.
 4. `Abort()` can be used to release the lease (`UPDATE RebusMessage SET LeasedUntilUtc = NULL WHERE... `)

The above is a bit of an oversimplification. There's two issues that need to be addressed; 
  1. What happens if a lease expires? Eg. If a worker is processing Message A and the machines loses power. This can be rectified by making the fetch step (simplified): `SELECT * FROM RebusMessage WHERE LeasedUntilUtc IS NULL OR LeasedUntilUtc < GETUTCDATE()` (Realistically you'd add some buffer in here)
  2. Now you're faced with the issue of what happens if a lease is for 45s but a worker takes 1m to process it. This can be fixed by having the implementation of `ITransactionContext` start a timer for, say, 1/2 the lease time. When the timer fires we extend the lease. Then if the worker machine loses power it'll stop renewing the lease and another worker can then process the message. As long as the handler is operating it'll periodically renew the lease.

This lease based approach is how, for instance, Azure Message Queues work. And I can see you even implement the auto renew lock logic for message queues (Eg. https://github.com/rebus-org/Rebus/wiki/Azure-Service-Bus-transport)

I'm yet to do any work on this as yet... but just wondering how you feel about this proposal? I don't think such an implementation would replace the current transactional semantics - which are useful especially in a local scenario - but more an extension/alternate implementation. Our specific use case is we process some large files that can take a a while to process. The nature of this processing means it's not possible to break down the process into smaller steps (Eg. process 1mb of the file at a time) due to internal state used in third party, closed source, libraries.

**Comment by mookid8000** [comment]
I have thought about implementing a lease-based transport for SQL Server for a while, and I think it makes good sense to do so.

I also believe that it makes good sense, performance-wise: I think it will be more scalable because it doesn't have to maintain a full connection with an ongoing transaction for the entire duration in which the message is handled.

A lease-based implementation could be made by simply extending the current schema, exactly as you propose.

It would not raise any problems that Rebus does not already handle somewhere else – the Azure Service Bus transport too has a lease-based protocol to implement the receive transaction.

I agree that the current transport should be left unmodified (at least for a while) – it would be pretty easy to create a new transport implementation alongside the current one.

**Comment by MrMDavidson** [comment]
I've actually had a quick crack at doing this. I'll need more of a play when I get back in to work tomorrow... a couple of notes though;
  1. I think there will need to be some modification of the `SqlServerTransport`; just to make it more suitable for code sharing between itself and `SqlServerLeasedTransport`. I've got a mostly working implementation as it stands but there's chunks that are copy and pasted between the two. Which obviously isn't great long term.
  2. My current implementation only uses leases for receiving. So sending is still performed as is. I've seen some scenarios where this causes deadlocks if a handler sends out a bunch of messages (not super realistic but definitely not something that should break things). I'm considering using a mailbox type approach for sending. Basically there'd be a copy of the `RebusMessage` table as it currently stands but with an additional `BatchId` `uniqueidentifier` column called `RebusMessageOutbox`. When you do a `bus.Send(...)` the `SqlServerLeasedTransport` would do simple insert into `RebusMessageOutbox` and generate a `BatchId` which gets stored in the `ITransactionContext`. Each `Send(..)` would perform its own connection management (ie. Send would close the connection and call `Complete` on it) but reuse the `BatchId`. However the `ITransactionContext` `OnAbort` call back would delete from `RebusMessageOutbox` where `BatchId` matches. And `OnCommit` would do;
```SQL
DELETE	RMO
OUTPUT	deleted.body,
		deleted.headers, -- And other columns
INTO	RebusMessage (body, headers)
FROM	RebusMessageOutbox RMO
WHERE	RMO.BatchId = @BatchId
```

To transfer the messages from the "outbox" to the normal message queue.

**Comment by MrMDavidson** [comment]
I'm an idiot... a simpler approach might be for `Send()` to simply store the message in memory on the `ITransactionContext` but not even attempt to put them in the DB. The `ITransactionContext` `OnCommit()` callback would do the actual insert. Having them temporarily in the database doesn't buy us anything.

**Comment by mookid8000** [comment]
You are absolutely right – not about being an idiot, 😁  but about simply storing messages in-mem until the queue transaction is committed. 

This is also how the ASB transport does it. Easy! 🍋

**Comment by MrMDavidson** [comment]
Haha, cheers. I'll do an implementation tomorrow and see how it goes!

**Comment by MrMDavidson** [comment]
Managed to find some time to quickly prototype in memory buffered sends. Works a treat. 👍 

Tomorrow I'll clean it up and commit it. Would you rather I open a PR against this repository? And if so; what'd be the time frame for the nuget package going out? Or would you rather I spin off a Rebus.SqlServer.Leased (or whatever) package?

**Comment by mookid8000** [comment]
let's just put it in Rebus.SqlServer

I am veeeeeery close to releasing Rebus 4.0.0, and when that is done I will go through all packages and release them.

Unfortunately I can't make any promises regarding the timeframe, but I am pretty confident that it will happen within the next 1-2-3 days or something.

**Comment by MrMDavidson** [comment]
Resolved with PR #14 being merged 🥂
